operator/internal/decommissioning/statefulset_decomissioner.go

```diff
@@ -0,0 +1,646 @@
+				continue
+			}
+
+			if strings.Contains(err.Error(), "does not exist") {
+				// delete the node from our sets
+				downedNodes.Delete(downedNode)
+				continue
+			}
+
+			log.Error(err, "fetching decommission status")
+			return false, err
+		}
+
+		if status.Finished {
+			// skip since we have already decommissioned it, so it should no longer
+			// show up in the health overview
+			continue
+		}
+
+		// add the brokers to the list of what needs to be decommissioned
+		currentlyDecommissioningBrokers = append(currentlyDecommissioningBrokers, downedNode)
+	}
+
+	sortBrokers := func(set []int) []int {
+		// sort by simple node id
+		sort.SliceStable(set, func(i, j int) bool {
+			return set[i] < set[j]
+		})
+		return set
+	}
+
+	healthyBrokers := sortBrokers(healthyNodes.Values())
+	brokersToDecommission = sortBrokers(brokersToDecommission)
+	brokersToIgnore = sortBrokers(brokersToIgnore)
+	currentlyDecommissioningBrokers = sortBrokers(currentlyDecommissioningBrokers)
+
+	formatBrokerList := func(set []int) string {
+		return strings.Join(functional.MapFn(strconv.Itoa, set), ", ")
+	}
+
+	log.V(traceLevel).Info(fmt.Sprintf(
+		"healthy brokers: [%s], ignored: [%s], to decommission: [%s], decommissioning: [%s]",
+		formatBrokerList(healthyBrokers),
+		formatBrokerList(brokersToIgnore),
+		formatBrokerList(brokersToDecommission),
+		formatBrokerList(currentlyDecommissioningBrokers),
+	))
+
+	if len(currentlyDecommissioningBrokers) != 0 {
+		// we skip decommissioning our next broker since we already have some node decommissioning in progress
+		return true, nil
+	}
+
+	if len(brokersToDecommission) > 0 {
+		// only record the event here since this is when we trigger a decommission
+		s.recorder.Eventf(set, corev1.EventTypeNormal, eventReasonBroker, "brokers needing decommissioning: [%s], decommissioning: %d", formatBrokerList(brokersToDecommission), brokersToDecommission[0])
+
+		if err := adminClient.DecommissionBroker(ctx, brokersToDecommission[0]); err != nil {
+			log.Error(err, "decommissioning broker", "broker", brokersToDecommission[0])
+			return false, err
+		}
+	}
+
+	// we should have decommissioned something above, so requeue and wait for it to finish
+	return true, nil
+}
+
+// findUnboundVolumeClaims fetches any PVCs associated with the StatefulSet that aren't actively attached
+// to a pod.
+//
+// Much of this code is copied from the original decommissioner and refactored, but the basic idea
+// is:
+//
+// 1. Pull any pods matching the labels for the stateful set's pod template that are in the same namespace
+// 2. Pull any pvcs matching the labels for the stateful set's volume claim template (though the component adds a "NAME-statefulset")
+// 3. Find unbound volumes by checking that the pods we pulled reference every volume claim
+//
+// NB: this follow the original implementation that has a potential race-condition in the cache, where a PVC may come online and be in-cache
```

Comment:
My preference would be to fix this race condition. If you delete a PVC and the pod schedules correctly, you'll get a zombie PVC. You can't reverse the deletion and the next time it comes unbound (e.i. the pod gets deleted for any reason) it'll get deleted.

I think you'd be safe if you instead filtered based on .status.phase and then used a deletion precondition?